frontend/src/ui/common.rs

```diff
@@ -25,6 +29,68 @@ pub const COLOR_REPS_RIR: usize = 4;
 pub const COLOR_WEIGHT: usize = 8;
 pub const COLOR_TIME: usize = 5;
 
+#[derive(Clone)]
+pub enum PlotType {
+    Circle(usize, u32),
+    Line(usize, u32),
+    Histogram(usize),
+}
+
+pub fn plot_line_with_dots(color: usize) -> Vec<PlotType> {
+    [PlotType::Line(color, 2), PlotType::Circle(color, 2)].to_vec()
+}
+
+#[derive(Default)]
+pub struct PlotParams {
+    pub y_min_opt: Option<f32>,
+    pub y_max_opt: Option<f32>,
+    pub secondary: bool,
+    pub omit_margin: bool,
+}
+
+impl PlotParams {
+    pub fn primary_range(min: f32, max: f32) -> Self {
+        Self {
+            y_min_opt: Some(min),
+            y_max_opt: Some(max),
+            secondary: false,
+            omit_margin: false,
+        }
+    }
+
+    pub const SECONDARY: Self = Self {
+        y_max_opt: None,
+        y_min_opt: None,
+        secondary: true,
+        omit_margin: false,
+    };
+}
+
+pub struct PlotData {
+    pub values: Vec<(NaiveDate, f32)>,
+    pub plots: Vec<PlotType>,
+    pub params: PlotParams,
+}
+
+#[derive(Clone, Copy, Default)]
+pub struct Bounds {
+    min: f32,
+    max: f32,
+    omit_y_margin: bool,
+}
+
+impl Bounds {
+    fn y_margin(self) -> f32 {
```

Comment:
For some reason there is no margin on the y axis in any of the charts anymore. Here is an example:

#### Now

![image](https://github.com/user-attachments/assets/acf854be-8b8c-4115-9c93-5949f5a0a55f)

#### Previously

![image](https://github.com/user-attachments/assets/cef8560d-31e7-432f-903b-e37c1bf66c2e)

Reply:
I mistakenly changed the multiplier at some point. This should be resolved now.

Reply:
The result still looks different. There shouldn't be a margin below 0. Such a margin gives the impression that the value could fall below 0.

#### Now

![image](https://github.com/user-attachments/assets/0c34c589-cd8b-4760-b4e5-8bfcaf10fd31)

#### `main` branch

![image](https://github.com/user-attachments/assets/7162a948-3b18-427d-aa10-8c02412b4895)

Reply:
I believe this is fixed now (frankly, I don't understand how/why this worked before...).

Reply:
Unfortunately, the behavior is still different. Previously, there was always a margin at the top. After the last change, this margin is removed if the chart starts at 0.

### Now

![image](https://github.com/user-attachments/assets/19884aaa-45d2-47ff-8090-1744d1b4f78f)

#### `main` branch

![image](https://github.com/user-attachments/assets/7162a948-3b18-427d-aa10-8c02412b4895)

Reply:
What's the intended behavior here?

Here is what I've implemented now:

- include no bottom margin if
   - margins is disable (has_margin=false)
   - carts starts at 0
- always include a top margin
- calculate the margin be 10% of the range of the y axis if this value can be calculated, 0.1 otherwise

If this is the intended semantics, I'd question the usefulness of the whole `has_margin` feature and would rather remove it altogether. The only place where this is used is the menstrual cycle chart, which starts at 0.0 and has the bottom margin disabled for that reason anyway.

Reply:
That looks good to me. I think you are right, the `has_margin` feature can be removed. I don't see where it would be needed.

Reply:
Removed.